File: sistema_asignacion.py
"""
Sistema Dinámico de Asignación - Opción 2 (reescritura clara)
Guía: Guía del Sistema Dinámico de Asignación.txt
4 factores: base + desplazamiento + flexibilidad + experiencia
"""
from dataclasses import dataclass
from enum import IntEnum


class Prioridad(IntEnum):
    P1 = 1
    P2 = 2
    P3 = 3


# Constantes según Guía
PUNTOS_BASE = {
    Prioridad.P1: 100,
    Prioridad.P2: 60,
    Prioridad.P3: 30,
}
BONO_DESPLAZAMIENTO = 100  # Guía.txt:9
BONO_FLEX_ULTIMA = 30      # Guía.txt:11 premio
PENALIZACION_FLEX_CON_RESPALDO = -15  # cede paso
MAX_EXP = 10               # Guía.txt:8


@dataclass(frozen=True)
class Solicitud:
    prioridad: Prioridad
    exp_anos: int = 0
    perdio_p1: bool = False
    perdio_p2: bool = False


@dataclass(frozen=True)
class ContextoNivel:
    total_opciones: int  # num materias en ese nivel de prioridad
    es_ultima: bool      # si es la última materia restante del nivel


def puntos_base(p: Prioridad) -> int:
    if p not in PUNTOS_BASE:
        raise ValueError(f"prioridad debe ser 1,2,3, recibido {p}")
    return PUNTOS_BASE[p]


def bono_desplazamiento(s: Solicitud, flex_prev_perdida: bool = False) -> int:
    """Guía.txt:9 protección + Nota 2 flex desplazado +30 si prev flexible perdida"""
    base = 0
    if s.prioridad == Prioridad.P2 and s.perdio_p1:
        base = BONO_DESPLAZAMIENTO
    if s.prioridad == Prioridad.P3 and (s.perdio_p1 or s.perdio_p2):
        base = BONO_DESPLAZAMIENTO
    flex_desplazado = BONO_FLEX_ULTIMA if flex_prev_perdida else 0
    return base + flex_desplazado


def ajuste_flexibilidad(c: ContextoNivel) -> int:
    """Guía.txt:10-11 independiente del bono, para cualquier nivel con >1 opciones"""
    if c.total_opciones <= 1:
        return 0
    return BONO_FLEX_ULTIMA if c.es_ultima else PENALIZACION_FLEX_CON_RESPALDO


def puntos_experiencia(anos: int) -> int:
    """Guía.txt:8 +1 por año, tope 10"""
    try:
        a = int(anos or 0)
    except (TypeError, ValueError):
        a = 0
    return min(max(a, 0), MAX_EXP)


def calcular_puntaje(solicitud: Solicitud, contexto: ContextoNivel, flex_prev_perdida: bool = False) -> int:
    """Suma 4 factores + flex desplazado intra-slot"""
    return (
        puntos_base(solicitud.prioridad)
        + bono_desplazamiento(solicitud, flex_prev_perdida)
        + ajuste_flexibilidad(contexto)
        + puntos_experiencia(solicitud.exp_anos)
    )


def calcular_puntaje_detalle(solicitud: Solicitud, contexto: ContextoNivel, flex_prev_perdida: bool = False) -> dict:
    """Útil para auditar / tests"""
    base = puntos_base(solicitud.prioridad)
    desp = bono_desplazamiento(solicitud, flex_prev_perdida)
    flex = ajuste_flexibilidad(contexto)
    exp = puntos_experiencia(solicitud.exp_anos)
    return {"base": base, "desplazamiento": desp, "flexibilidad": flex, "experiencia": exp, "total": base+desp+flex+exp, "flex_prev_perdida": flex_prev_perdida}


if __name__ == "__main__":
    # Demo
    casos = [
        (Solicitud(Prioridad.P1, 5, False, False), ContextoNivel(3, False)),
        (Solicitud(Prioridad.P2, 5, True, False), ContextoNivel(3, False)),
        (Solicitud(Prioridad.P2, 5, True, False), ContextoNivel(3, True)),
        (Solicitud(Prioridad.P3, 5, True, True), ContextoNivel(2, True)),
    ]
    for s, c in casos:
        print(s, c, "->", calcular_puntaje_detalle(s, c))

File: src/lib/sistemaAsignacion.ts
export enum Prioridad {
  P1 = 1,
  P2 = 2,
  P3 = 3,
}

export const PUNTOS_BASE: Record<Prioridad, number> = {
  [Prioridad.P1]: 100,
  [Prioridad.P2]: 60,
  [Prioridad.P3]: 30,
};

export const BONO_DESPLAZAMIENTO = 40; // Nivelador P2: 60 + 40 = 100 (mismo piso que P1)
export const BONO_DESPLAZAMIENTO_P3 = 70; // Nivelador P3: 30 + 70 = 100 si perdió P1/P2
export const BONO_FLEX_ULTIMA = 30;
export const PENALIZACION_FLEX_CON_RESPALDO = -15;
export const MAX_EXP = 8; // Máximo 8 puntos (1 punto por cada periodo impartido en los últimos 4 años)

export type Solicitud = {
  prioridad: Prioridad;
  expPeriodos: number; // Número de periodos impartidos en los últimos 4 años
  expAnos?: number; // Compatibilidad legacy
  perdioP1: boolean;
  perdioP2: boolean;
};

export type ContextoNivel = {
  totalOpciones: number;
  esUltima: boolean;
};

export function puntosBase(p: Prioridad): number {
  const v = PUNTOS_BASE[p];
  if (v === undefined) throw new Error(`prioridad debe ser 1,2,3, recibido ${p}`);
  return v;
}

export function bonoDesplazamiento(s: Solicitud, flexPrevPerdida = false): number {
  let base = 0;
  if (s.prioridad === Prioridad.P2 && s.perdioP1) base = BONO_DESPLAZAMIENTO;
  if (s.prioridad === Prioridad.P3 && (s.perdioP1 || s.perdioP2)) base = BONO_DESPLAZAMIENTO_P3;
  const flexDesplazado = flexPrevPerdida ? BONO_FLEX_ULTIMA : 0;
  return base + flexDesplazado;
}

export function ajusteFlexibilidad(c: ContextoNivel): number {
  if (c.totalOpciones <= 1) return 0;
  return c.esUltima ? BONO_FLEX_ULTIMA : PENALIZACION_FLEX_CON_RESPALDO;
}

export function puntosExperiencia(periodos: number): number {
  const p = Number(periodos ?? 0);
  if (Number.isNaN(p)) return 0;
  // +1 punto por cada periodo impartido en los últimos 4 años, tope MAX_EXP (8 puntos)
  return Math.min(Math.max(Math.trunc(p), 0), MAX_EXP);
}

export function calcularPuntaje(solicitud: Solicitud, contexto: ContextoNivel, flexPrevPerdida = false): number {
  const expVal = solicitud.expPeriodos ?? solicitud.expAnos ?? 0;
  return (
    puntosBase(solicitud.prioridad) +
    bonoDesplazamiento(solicitud, flexPrevPerdida) +
    ajusteFlexibilidad(contexto) +
    puntosExperiencia(expVal)
  );
}

export function calcularPuntajeDetalle(solicitud: Solicitud, contexto: ContextoNivel, flexPrevPerdida = false) {
  const expVal = solicitud.expPeriodos ?? solicitud.expAnos ?? 0;
  const base = puntosBase(solicitud.prioridad);
  const desp = bonoDesplazamiento(solicitud, flexPrevPerdida);
  const flex = ajusteFlexibilidad(contexto);
  const exp = puntosExperiencia(expVal);
  return { base, desplazamiento: desp, flexibilidad: flex, experiencia: exp, periodos: expVal, total: base + desp + flex + exp, flexPrevPerdida };
}
